src/Grpc.Net.Client.Web/Internal/GrpcWebResponseStream.cs

```diff
@@ -52,59 +58,113 @@ public override async ValueTask<int> ReadAsync(Memory<byte> data, CancellationTo
 #if NETSTANDARD2_0
         var data = buffer.AsMemory(offset, count);
 #endif
+        Memory<byte> headerBuffer = Memory<byte>.Empty;
 
         switch (_state)
         {
             case ResponseState.Ready:
-                // Read the header first
-                // - 1 byte flag for compression
-                // - 4 bytes for the content length
-                Memory<byte> headerBuffer;
-
-                if (data.Length >= 5)
                 {
-                    headerBuffer = data.Slice(0, 5);
+                    if (data.Length == 0)
+                    {
+                        // Handle zero byte reads.
+                        var read = await StreamHelpers.ReadAsync(_inner, data, cancellationToken).ConfigureAwait(false);
+                        Debug.Assert(read == 0);
+                        return 0;
+                    }
+                    else
+                    {
+                        // Read the header first
+                        // - 1 byte flag for compression
+                        // - 4 bytes for the content length
+                        _contentRemaining = HeaderLength;
+                        _state = ResponseState.Header;
+                        goto case ResponseState.Header;
+                    }
                 }
-                else
+            case ResponseState.Header:
                 {
-                    // Should never get here. Client always passes 5 to read the header.
-                    throw new InvalidOperationException("Buffer is not large enough for header");
+                    headerBuffer = data.Length >= _contentRemaining ? data.Slice(0, _contentRemaining) : data;
+                    var success = await TryReadDataAsync(_inner, headerBuffer, cancellationToken).ConfigureAwait(false);
+                    if (!success || headerBuffer.Length == 0)
+                    {
+                        return 0;
```

Comment:
Can `_contentRemaining` ever be zero here? If so, this seems like it could incorrectly report the end of stream. It looks like this might be impossible because you should transition to the `Ready` state first, so maybe just add an assert?

```suggestion
                        Debug.Assert(_contentRemaining > 0);
                        return 0;
```